### PR TITLE
🚧 Add basic auth setup

### LINKED ISSUE #54: Add Authentification

### PR DESCRIPTION
**Pull request by mhstrkmp**
## Description

- Add basic authentication setup with useContext api
- Following [this tutorial by Kent C. Dodds](https://kentcdodds.com/blog/authentication-in-react-applications) as it seems to be an elegant solution
- Login and logout functionality follows ...

## Related Issue
- closes #54

### CODE REVIEW

File: client/src/App.js [hunk 1]
```diff
@@ -1,59 +1,11 @@
...
-              <ProfilePage title={"Profil"} />
-            </Route>
-          </Switch>
-        </Router>
-        <ReactQueryDevtools initialIsOpen={false} />
-      </QueryClientProvider>
-    </>
-  );
+  const user = useUser();
+  return user ? <AuthenticatedApp /> : <UnauthenticatedApp />;
```
Comment (boldCM) [inline client/src/App.js:8]: I like that! 🎉

File: client/src/AuthenticatedApp.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import { ProfilePage } from "./pages/ProfilePage/ProfilePage";
+import { QueryClient, QueryClientProvider } from "react-query";
+import { ReactQueryDevtools } from "react-query/devtools";
+import { useAuth } from "./context/AuthContext";
+
+const queryClient = new QueryClient();
+
+function AuthenticatedApp() {
+  const data = useAuth();
+  console.log(data);
```
Comment (boldCM) [inline client/src/AuthenticatedApp.js:20]: do you still need that console.log? ;)

File: server.js [hunk 3]
```diff
@@ -97,6 +98,22 @@ app.route("/api/items/:id").get(async (req, res) => {
...
+  // Send username, password
+  const { username, password } = req.params;
+  try {
+    // Search for corresponding user
+    const userData = await authenticateUser("users", username, password);
+    res.send(userData);
+  } catch (error) {
+    console.error(error);
+    res
+      // Error message: Wrong username or password please try again
```
Comment (boldCM) [inline server.js:111]: maybe put that errore message in a res.send? ;)

**Review by boldCM — APPROVED** [review]
lgtm so far 🎉
ahevn't worked with authentification yet, if you need specfic reviews for that maybe ask for another review ;)

File: client/src/App.js (hunk 1 above)
Comment (mhstrkmp) [reply client/src/App.js:8]: Yeah me too. It seems to bee an elegant solution.

File: client/src/context/UserContext.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import React, { createContext, useContext } from "react";
+import { useAuth } from "./AuthContext";
+
+const UserContext = createContext();
+const UserProvider = (props) => (
+  <UserContext.Provider value={useAuth().data.user} {...props} />
```
Comment (ncbecker) [inline client/src/context/UserContext.js:6]: Do you need a separate context as it just provides a value that could already provided by AuthContext?

Maybe you could add this function to your AuthContext.js:
```
export function useUser() {
  return useAuth().data.user;
}
```

**Review by ncbecker — APPROVED** [review]
Well done, great code quality as always 👍

File: client/src/context/AuthContext.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import React, { createContext, useContext, useEffect, useState } from "react";
+import axios from "axios";
```
[the remark this answers is not in the document]
Comment (mhstrkmp) [reply client/src/context/AuthContext.js:2]: Good point. Think I just wanted to play 😄

File: client/src/context/AuthContext.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+const AuthContext = createContext();
+
+function AuthProvider(props) {
+  const [data, setData] = useState({});
```
[the remark this answers is not in the document]
Comment (mhstrkmp) [reply client/src/context/AuthContext.js:7]: It will get more or different data later therefore a more generic naming

**Pull request merged**